src/xpmem.c

```diff
@@ -0,0 +1,219 @@
+	 * fail. To prevent this we need to determine the value of
+	 * TASK_SIZE. On x86_64 the value was hard-coded in sm to be
+	 * 0x7ffffffffffful but this approach does not work with AARCH64
+	 * (and possibly other architectures). Since there is really no
+	 * way to directly determine the value we can (in all cases?) look
+	 * through the mapping for this process to determine what the
+	 * largest address is.  This should be the top of the stack. No
+	 * heap allocations should be larger than this value.  Since the
+	 * largest address may differ between processes the value must be
+	 * shared as part of the modex and stored in the endpoint. */
+	int ret = 0;
+	char buffer[1024];
+	uintptr_t address_max = 0;
+	FILE *fh;
+
+	fi_param_define(&core_prov, "use_xpmem", FI_PARAM_BOOL,
+			"Whether to use XPMEM over CMA when possible "
+			"(default: no)");
+	fi_param_define(&core_prov, "xpmem_memcpy_chunksize", FI_PARAM_SIZE_T,
+			"Maximum size used for a single memcpy call"
+			"(default: %d)", XPMEM_DEFAULT_MEMCPY_CHUNK_SIZE);
+
+	xpmem = calloc(sizeof(*xpmem), 1);
+	if (!xpmem)
+		return -FI_ENOMEM;
+
+	fh = fopen("/proc/self/maps", "r");
+	if (NULL == fh) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+				"could not open /proc/self/maps for reading");
+		ret = -FI_EIO;
+		goto fail;
+	}
+
+	while (fgets(buffer, sizeof(buffer), fh)) {
+		uintptr_t low, high;
+		char *tmp;
+		/* each line of /proc/self/maps starts with low-high in
+		 * hexidecimal (without a 0x) */
+		low = strtoul(buffer, &tmp, 16);
+		high = strtoul(tmp + 1, NULL, 16);
+		if (address_max < high) {
+			address_max = high;
+		}
+	}
+	fclose(fh);
+
+	if (address_max == 0) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"could not determine the address max");
+		ret = -FI_EIO;
+		goto fail;
+	}
+
+	/* save the calcuated maximum */
+	xpmem->pinfo.address_max = address_max - 1;
+
+	/* export the process virtual address space for use with xpmem */
+	xpmem->pinfo.seg_id = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE,
+					 (void *) 0666);
+	if (xpmem->pinfo.seg_id == -1) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to export process virtual address space for use with xpmem\n");
+		ret = -FI_ENODATA;
+		goto fail;
+	}
+
+	ret = fi_param_get_size_t(&core_prov, "xpmem_memcpy_chunksize",
+				  &xpmem->memcpy_chunk_size);
+	if (ret)
+		xpmem->memcpy_chunk_size = XPMEM_DEFAULT_MEMCPY_CHUNK_SIZE;
+
+	return 0;
+
+fail:
+	free(xpmem);
+	xpmem = NULL;
+	return ret;
+}
+
+int xpmem_cleanup(void)
+{
+	int ret = 0;
+
+	if (xpmem)
+		ret = xpmem_remove(xpmem->pinfo.seg_id);
+
+	return (ret) ? -FI_EINVAL : ret;
+}
+
+static inline void xpmem_memcpy(void *dst, void *src, size_t size)
+{
+	while (size > 0) {
+		size_t copy_size = MIN(size, xpmem->memcpy_chunk_size);
+		memcpy(dst, src, copy_size);
+		dst = (void *) ((uintptr_t) dst + copy_size);
+		src = (void *) ((uintptr_t) src + copy_size);
+		size -= copy_size;
+	}
+}
+
+int xpmem_copy(struct ofi_mr_cache *cache, struct iovec *local,
```

Comment:
it feels like this cache could be internal to the xpmem code instead of having to pass it into the copy function. from what I can tell, it gets initialized and destroyed by shm but all other searches/cache delete calls are done inside xpmem

Reply:
When you receive a request to transfer memory using XPMEM,  you need to cache the address entry to avoid having to do xpmem_attach()/xpmem_detach() repeatedly. The cache entry needs to be keyed by peer_id. Because the virtual memory address space of different processes would overlap. if the cache is global, would you run into a problem then? Say you have multiple domains/endpoints using XPMEM?

Reply:
As one option, you could include the peer id as part of the cache search/hit criteria.  Are you creating 1 cache per peer? Or just a single cache per shm domain?

Reply:
@shefty That's what's happening [right now](https://github.com/ofiwg/libfabric/blob/main/prov/util/src/util_mr_cache.c#L59-L62)
There is one cache per domain which is a peer set/set of peer ids
Just spit balling but maybe we could have one internal global cache (inside of the xpmem code) and pass in the pid instead?

Reply:
@aingerson, @shefty, I'm not in favor of using a global cache. Even if we use the pid, the pid will be the same globally, and you might hit duplicate cache entries if you have multiple domains talking to the same processes. I think having SHM (or other providers) initialize the cache per domain is the cleanest way to handle this issue.

Reply:
Would it be problematic to have it overlap between domains? The domains are probably using different memory regions but even if they are using the same memory, any duplicates should get caught by the cache and help avoid extra calls to xpmem_attach.

Reply:
While I understand the idea behind the comment, I'm having trouble justifying the difference in cache usage between IPC and XPMEM. IPC cache exists within the domain. Why not have a global cache in that case also? Is there a functional reason IPC cache needs to be within a domain? Would it be more consistent to have both IPC and XPMEM cache be created in the same way?

Reply:
The IPC cache and XPMEM cache are super similar and can be used in the same way. The idea behind putting it in the domain is because shm is directly calling into it before and after the copy instead of passing it into the copy call. Not all interfaces support or need the cache so it's interface dependent. It's up to shm to use the cache how appropriate:
```
	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ZE) {
		id = cmd->msg.hdr.id;
		ipc_device = cmd->msg.data.ipc_info.device;
		fd = ep->sock_info->peers[id].device_fds[ipc_device];
		ret = ze_hmem_open_shared_handle(fd,
				(void **) &cmd->msg.data.ipc_info.ipc_handle,
				&ipc_fd, ipc_device, &base);
	} else {
		ret = ofi_ipc_cache_search(domain->ipc_cache,
					   cmd->msg.hdr.id,
					   &cmd->msg.data.ipc_info,
					   &mr_entry);
	}
```
We could make it global to each interface which is also fine but it would also require passing in peer identifying information. If you don't want to make it global, I would suggest leaving it in the domain but calling it around the vma copy call instead of passing it in so that it matches the ipc cache usage. Would that be better for you?

Reply:
As discussed I made domain optional in mr cache initialization. And moved xpmem cache to global. Now I use PID as the global identifier. Now pid is used for both CMA and XPMEM